vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,625 @@
+            size: counter,
+            reserved: [0; 5],
+            val: cfg,
+        })
+    }
+
+    pub fn read_name_config(&self) -> Result<VuInputConfig> {
+        let mut name: [u8; VIRTIO_INPUT_CFG_SIZE] = [0; VIRTIO_INPUT_CFG_SIZE];
+
+        // SAFETY: Safe as the file is a valid event device, the kernel will only
+        // update the correct amount of memory in func.
+        match unsafe { eviocgname(self.ev_dev.get_raw_fd(), name.as_mut_slice()) } {
+            Ok(len) if len as usize > name.len() => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            Ok(len) if len <= 1 => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            Err(_) => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            _ => (),
+        }
+
+        let size = String::from_utf8(name.to_vec()).unwrap().len();
+
+        Ok(VuInputConfig {
+            select: self.select,
+            subsel: 0,
+            size: size as u8,
+            reserved: [0; 5],
+            val: name,
+        })
+    }
+
+    pub fn read_id_config(&self) -> Result<VuInputConfig> {
+        let input_id = self.ev_dev.input_id();
+
+        let mut dev_id = [
+            input_id.bus_type().0.as_slice(),
+            input_id.vendor().as_slice(),
+            input_id.product().as_slice(),
+            input_id.version().as_slice(),
+        ]
+        .concat();
+
+        dev_id.resize(VIRTIO_INPUT_CFG_SIZE, 0);
+
+        Ok(VuInputConfig {
+            select: VIRTIO_INPUT_CFG_ID_DEVIDS,
+            subsel: 0,
+            size: VIRTIO_INPUT_CFG_SIZE as u8,
+            reserved: [0; 5],
+            val: dev_id.try_into().unwrap(),
+        })
+    }
+}
+
+/// VhostUserBackend trait methods
+impl<T: 'static + InputDevice + Sync + Send> VhostUserBackendMut<VringRwLock, ()>
+    for VuInputBackend<T>
+{
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_RING_F_INDIRECT_DESC
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ | VhostUserProtocolFeatures::CONFIG
+    }
+
+    fn get_config(&self, _offset: u32, size: u32) -> Vec<u8> {
+        let cfg = match self.select {
+            VIRTIO_INPUT_CFG_ID_NAME => self.read_name_config(),
+            VIRTIO_INPUT_CFG_ID_DEVIDS => self.read_id_config(),
+            VIRTIO_INPUT_CFG_EV_BITS => self.read_event_config(),
+            _ => {
+                // Return zeros for unsupported config types
+                return vec![0; size as usize];
+            }
+        };
+
+        match cfg {
+            Ok(val) => val.to_bytes(),
```

Comment:
This needs to be offsetted and sliced according to the function parameters. Ideally, we move this into the vhost crate. I was about to point you at vhu_scisi, but realized that it is broken there too... So instead you can look here:
https://gitlab.com/virtio-fs/virtiofsd/-/blob/e6545e1002f1bda486aaa676b1f608d0eb78e9c2/src/main.rs#L456-464

Reply:
Has fixed it.  Thanks for shared pointer.